### PR TITLE
enable average tokens across devices

### LINKED ISSUE #34242: Add DDP token averaging for equivalent non-parallel training similar to #34191
### Feature request

Token averaging in gradient accumulation was fixed in #34191 . But token averaging in DDP seems to have the same issue.

---

## Expected behaivor

With all the tokens contributing to loss in each step (in each GPU, gradient accumulation step, and microbatch), the equation becomes:

$$ntokens=\sum\limits_{GPUs} \sum\limits_{gas} \sum\limits_{microb} (label\neq-100)$$

I believe we should average the above tokens at the same time for equivalent non-parallel training.

---

## Current issue

Prior to #34191, the loss/gradients were averaged on $\sum\limits_{GPUs}$, $\sum\limits_{gas}$, and $\sum\limits_{microb}$ separately. And, the introduction of `num_items_in_batch` in #34191 refers to:

$$ntokens=\sum\limits_{gas} \sum\limits_{microb} (label\neq-100)$$

So, the loss/gradients are now averaged on $\sum\limits_{GPUs}$ and $\left(\sum\limits_{gas}\sum\limits_{microb}\right)$ separately. However, this still does not seem equivalent to non-parallel training.

Can we also incorporate $\sum\limits_{GPUs}$ when determining `num_items_in_batch`? Something like `all_reduce(num_items_in_batch)`?

### Motivation

DDP seems not fully equivalent to non-parallel training.

related comments: https://github.com/huggingface/transformers/pull/34191#issuecomment-2421777304

### Your contribution

Found some fairseq implementation of this feature

https://github.com/facebookresearch/fairseq/blob/018621f3cca02ca9de945dc082c3fb1a7f9f2deb/fairseq/trainer.py#L932-L949

### PR DESCRIPTION
**Pull request by techkang**
# What does this PR do?

Fix [https://github.com/huggingface/transformers/issues/34242](https://github.com/huggingface/transformers/issues/34242)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@muellerzr pls help to check.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3636,7 +3636,15 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
...
+                        self.args.world_size > 1):
+                    num_items_in_batch_tensor = torch.tensor(num_items_in_batch, device=self.args.device)
+                    num_items_in_batch = int(self.accelerator.gather(num_items_in_batch_tensor).sum().cpu())
+                    device_count_for_loss = self.args.world_size
+                else:
+                    device_count_for_loss = 1
                 loss = self.compute_loss_func(outputs, labels, num_items_in_batch=num_items_in_batch)
+                loss *= device_count_for_loss
             elif model_name in MODEL_FOR_CAUSAL_LM_MAPPING_NAMES.values():
                 loss = self.label_smoother(outputs, labels, shift_labels=True)
```
Comment (muellerzr) [inline src/transformers/trainer.py:3649]: Thanks! We also need the case for when we don't define this, e.g. it's passed to the model `forward()`. So what would be better is to perform the `gather` much earlier, and pass the new `num_items_in_batch` as part of the call to `compute_loss`. 

And then perform the `loss *=` where we call `loss *= self.args.gradient_accumulation_steps` later (right before we call `accelerator.backward()`)

**Review by muellerzr — COMMENTED** [review]
Thanks! It's a good start, let's improve it a little :)

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -1530,6 +1530,13 @@ class TrainingArguments:
         },
     )
 
+    average_tokens_across_devices: Optional[bool] = field(
+        default=False,
+        metadata={
+            "help": "Whether or not to average tokens across devices."
+        }
+    )
```
Comment (muellerzr) [inline src/transformers/training_args.py:1538]: During the `__post_init__` we call `setup_devices`. We can change `average_tokens_across_devices` value to `False` if the world size < 1 I think!

This then simplifies it earlier to just be `if self.args.average_tokens_across_devices`

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -3603,6 +3603,9 @@ def training_step(
                 scaled_loss.backward()
         else:
             loss *= self.args.gradient_accumulation_steps
+            if (self.args.average_tokens_across_devices and num_items_in_batch is not None and
+                    self.args.world_size > 1):
+                loss *= self.args.world_size
```
Comment (muellerzr) [inline src/transformers/trainer.py:3608]: Both of these chunks I think can be under a `if num_items_in_batch is not None`, since both need to be valid for the `loss *= self.args.gradient_accumulation_steps`

**Review by muellerzr — COMMENTED** [review]
Thanks! We're quite close. Two more nits

File: src/transformers/training_args.py (hunk 2 above)
Comment (techkang) [reply src/transformers/training_args.py:1538]: Thanks, already fixed it, please check.

File: src/transformers/trainer.py (hunk 3 above)
Comment (techkang) [reply src/transformers/trainer.py]: I think gradient accumulation is orthogonal to DDP, and used a new if statement. Please check my code. Thanks.

Comment (muellerzr) [reply src/transformers/trainer.py]: It is, it's a matter of `self.model.accepts_loss_kwargs`

**Review by muellerzr — APPROVED** [review]
This indeed looks like the correct solution! Thanks!

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -1530,6 +1530,13 @@ class TrainingArguments:
         },
     )
 
+    average_tokens_across_devices: Optional[bool] = field(
+        default=False,
+        metadata={
+            "help": "Whether or not to average tokens across devices."
```
Comment (SunMarc) [inline src/transformers/training_args.py:1536]: Maybe we could share a bit more why this arg could be useful ?

**Review by SunMarc — APPROVED** [review]
Thanks for your work ! LGTM !

File: src/transformers/training_args.py (hunk 4 above)
Comment (techkang) [reply src/transformers/training_args.py]: Done, please review my code.

**Comment by techkang** [comment]
@muellerzr @SunMarc I fixed one typo to make ruff happy and pass auto check. And added more explanation to the argument. Pls check again, thank you.

**Comment by techkang** [comment]
One required tests failed but it's due to timeout. Can any one help to fix it? Thanks a lot.

**Comment by techkang** [comment]
I found some merged PRs encountered the same CI failure. So I don't think it's not caused by this PR.
https://github.com/huggingface/transformers/pull/34377/checks?check_run_id=32020760435

**Review by muellerzr — APPROVED** [review]
Thanks! Overall LG2M, cc @SunMarc

**Comment by muellerzr** [comment]
We'll merge once our CI allows us to 😅

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -3602,7 +3602,12 @@ def training_step(
...
-            loss *= self.args.gradient_accumulation_steps
+            if num_items_in_batch is not None:
+                if self.compute_loss_func or self.model_accepts_loss_kwargs:
+                    loss *= self.args.gradient_accumulation_steps
+                # Average tokens across devices is orthogonal to gradient accumulation
+                if self.args.average_tokens_across_devices:
+                    loss *= self.args.world_size
             self.accelerator.backward(loss, **kwargs)
 
         return loss.detach() / self.args.gradient_accumulation_steps
```
Comment (SunMarc) [inline src/transformers/trainer.py:3613]: We modified a lot how loss is computed, are we sure that this is loss is the same as the one applied ?

**Review by SunMarc — APPROVED** [review]
Thanks ! LGTM ! Just a nit

**Comment by muellerzr** [comment]
cc @ydshieh

**Review by ArthurZucker — APPROVED** [review]
Thanks! Rebase to make sure we have green CIs, will merge anyways!

**Comment by ArthurZucker** [comment]
Thanks @techkang 🤗
